### PR TITLE
[bf16] add bf16 support for fused adam

### PR DESCRIPTION
**Pull request by ver217**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#3839 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

1. Add bf16 support for fused adam optimizer.
2. Update related tests.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_optimizer/test_fused_adam_kernel.py [hunk 1]
```diff
@@ -41,11 +43,9 @@ def torch_adam_update(
     param.addcdiv_(exp_avg, denom, value=-step_size)
 
 
-@clear_cache_before_run()
```
Comment (FrankLeeeee) [inline tests/test_optimizer/test_fused_adam_kernel.py:44]: you can keep this to save GPU memory.

Comment (ver217) [reply tests/test_optimizer/test_fused_adam_kernel.py:44]: This is not compatible with `@pytest.mark.parametrize`

Comment (ver217) [reply tests/test_optimizer/test_fused_adam_kernel.py:44]: image
This leads to 
image

Comment (FrankLeeeee) [reply tests/test_optimizer/test_fused_adam_kernel.py:44]: Right, it is only compatible with `colossalai.testing.parameterize`.

Comment (ver217) [reply tests/test_optimizer/test_fused_adam_kernel.py:44]: But it seems that `colossalai.testing.parameterize` does not support usage like 
image

Comment (FrankLeeeee) [reply tests/test_optimizer/test_fused_adam_kernel.py:44]: I see, that's true.

**Pull request merged**